File: mysql_client/cpp/conn_mysql.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <thread>
#include <mutex>
#include <condition_variable>

#include <driver/mysql_driver.h>
#include <driver/mysql_connection.h>

#include <cppconn/driver.h>
#include <cppconn/exception.h>
#include <cppconn/statement.h>
#include <cppconn/resultset.h>

#include <stdlib.h>
#include <time.h>
#include <math.h>

#include "mysql_config.hpp"

using namespace std;

const int ROW = 1*1000*1000;
const int THR = 10;
const int BAR = 50;

int GRA = (int)sqrt(sqrt(ROW));

mutex mtx;
condition_variable cv;

string rs(int n){
	ostringstream o;
	for(int i=0;i<n;i++){
		int r = rand()%26+65;
		o << (char)r;
	}
	return o.str();
}

void do_query(sql::Driver *driver, int s, int e, int* c){

	sql::Connection *con;
	sql::Statement *stmt;

	con = driver->connect(HOST, USER, PASSWORD);
	con->setSchema(DB);

	stmt = con->createStatement();

	int t = 0;
	for(int i=s;i<e;i++){
		ostringstream sql;
		sql << "INSERT INTO t2 (col2, col3) VALUES (" << i << ", '" << rs(32) << "')";
		stmt->execute(sql.str());
		if(i%GRA == 0){
			mtx.lock();
			*c += i-s-t;
			cv.notify_all();
			mtx.unlock();
			t = i-s;
		}
	}

	if(t<(e-s)){
		mtx.lock();
		*c += e-s-t;
		cv.notify_all();
		mtx.unlock();
	}

	delete stmt;
	delete con;

}

void do_count(int* c, time_t start){

	unique_lock<mutex> lk(mtx);
	time_t end;

	while(1){
		cv.wait(lk);
		cout << "[";

		for(int i=0; i< ((*c)*100/ROW)/(100/BAR); i++ ){
			cout << "#";
		}

		for(int i=0; i< BAR-((*c)*100/ROW)/(100/BAR); i++ ){
			cout << " ";
		}

		end = time(NULL);
		cout << "] " << *c << " [" << (*c)*100/ROW << "%] (/" << ROW << ") : " << end-start << "s " << *c/(end-start==0?1:end-start) << "qps\r" << flush;

		if(ROW<=*c) break;
	}

	cout << endl;
}

int main() {
	cout << HOST << ", " << USER << ", " << PASSWORD << ", " << DB << endl;

	//sql::ResultSet *res;

	srand(time(NULL));

	try {
		sql::Driver *driver;
		sql::Connection *con;

		driver = sql::mysql::get_driver_instance();
		con = driver->connect(HOST, USER, PASSWORD);
		con->setSchema(DB);

		sql::Statement *stmt;

		stmt = con->createStatement();
		stmt->execute("DROP TABLE IF EXISTS t2");
		stmt->execute("CREATE TABLE t2 (col1 integer NOT NULL PRIMARY KEY AUTO_INCREMENT, col2 integer, col3 varchar(256))");

		delete stmt;

		time_t start = time(NULL);
		vector<thread> threads;

		int c = 0;
		for(int i=0;i<THR;i++){
			int s = ROW/THR*i;
			int e = ROW/THR*(i+1);
			threads.push_back(thread(do_query,driver,s,e,&c));
		}

		threads.push_back(thread(do_count,&c,start));

		for(int i=0;i<THR+1;i++){
			threads[i].join();
		}

		delete con;

	} catch (sql::SQLException &e) {
		cout << e.what() << endl;
	} catch (runtime_error &e) {
		cout << e.what() << endl;
	}

	return 0;
}

File: mysql_client/go/conn_mysql.go
package main

import (
	"database/sql"
	"fmt"
	"math/rand"
	"math"
	"time"

	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"


	. "./config"
)

const (
	ROW = 1*1000*1000
	THR = 10
	BAR = 50
	TBL = "t2"
)

var (
	GRA = int(math.Sqrt(ROW)*10)
)

func main() {

	dsn := USER+":"+PASS+"@tcp("+HOST+":3306)/"+DB

	db, err := sql.Open("mysql", dsn)
	if err != nil {
		panic(err.Error())
	}

	fmt.Println(dsn,TBL)

	_, _ = db.Exec("DROP TABLE "+TBL)

	_, err = db.Exec("CREATE TABLE "+TBL+" (col1 integer NOT NULL PRIMARY KEY AUTO_INCREMENT, col2 integer, col3 varchar(256))")
	if err != nil {
		panic(err.Error())
	}

	db.Close()

	finalize := make(chan string)
	report := make(chan int)
	for i:=0; i < THR; i++ {
		go exec(dsn,ROW/THR*i,ROW/THR*(i+1),i,finalize,report)
	}

	go do_report(finalize,report)

	m := 0 
	for {
		_ = <- finalize
		m += 1
		if m == THR+1 { break }
	}
	fmt.Println("\nFinished.")

}

func exec(dsn string, s,e,id int, finalize chan string, report chan int){

	db, err := sql.Open("mysql", dsn)
	if err != nil {
		panic(err.Error())
	}
	defer db.Close()

	t := 0
	for i:=s; i < e; i++ {
		_, err := db.Exec(fmt.Sprintf("INSERT INTO "+TBL+" (col2, col3) VALUES (%d, '%s')",i,rsa(32)))
		if err != nil {
			panic(err.Error())
		}
		if i % ((e-s)/GRA) == 0 {
			report <- i-s-t
			t = i-s
		}
	}
	if (e-s)%GRA != 0 {
		report <- e-s-t
	}
	finalize <- "Finished."

}

func rs() string {
	rand.Seed(time.Now().UnixNano())
	return string(rand.Int63()%26+65)
}

func rsa(w int) string {
	s := ""
	for i:=0; i<w; i++ {
		s += rs()
	}
	return s
}

func do_report(finalize chan string, report chan int){
	s := time.Now()

	n := 0
	pn := 0
	for {
		n += <- report

		if ROW/GRA <= n {
			pn += n
			n = 0
			disp(pn,s)

		}

		if ROW <= pn+n {
			disp(pn+n,s)
			break
		}
	}
	finalize <- "Finished."
}

func disp(pn int, s time.Time){
	e := int(time.Since(s)/(1000*1000*1000))

	fmt.Printf("\033[2K\033[G[");
	for i:=0; i< (pn*100/ROW)/(100/BAR); i++ {
		fmt.Printf("#");
	}
	for i:=0; i< BAR-(pn*100/ROW)/(100/BAR); i++ {
		fmt.Printf(" ");
	}
	if(0<e){
		fmt.Printf("] %d [%d%%] (/%d) : %ds : %d qps",pn,(pn*100/ROW),ROW,e,pn/e)
	}

}
